http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java

```diff
@@ -1491,182 +1449,359 @@ private <I> Publisher<URI> resolveURI(io.micronaut.http.HttpRequest<I> request,
+     */
+    private Mono<NettyClientByteBodyResponse> sendRawRequest(
+        ConnectionManager.PoolHandle poolHandle,
+        io.micronaut.http.HttpRequest<?> request,
+        NettyByteBody byteBody,
+        UnaryOperator<Throwable> mapConnectionError
+    ) {
+        URI uri = request.getUri();
+        String uriWithoutHost = uri.getRawPath();
+        if (uri.getRawQuery() != null) {
+            uriWithoutHost += "?" + uri.getRawQuery();
+        }
+        HttpRequest nettyRequest = NettyHttpRequestBuilder.asBuilder(request)
+            .toHttpRequestWithoutBody()
+            .setUri(uriWithoutHost);
+
+        return Mono.<NettyClientByteBodyResponse>create(sink -> {
+            if (log.isDebugEnabled()) {
+                log.debug("Sending HTTP {} to {}", request.getMethodName(), request.getUri());
+            }
 
-        // build the mock netty request (only the content-type matters)
-        HttpRequest nettyRequest = new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/");
-        List<AsciiString> relevantHeaders = List.of(HttpHeaderNames.CONTENT_TYPE);
-        for (AsciiString header : relevantHeaders) {
-            nettyRequest.headers().add(header, request.getHeaders().getAll(header));
-        }
+            boolean expectContinue = HttpUtil.is100ContinueExpected(nettyRequest);
+            ChannelPipeline pipeline = poolHandle.channel.pipeline();
 
-        HttpPostRequestEncoder encoder = buildMethod.apply(nettyRequest);
-        HttpRequest finalized = encoder.finalizeRequest();
-        // copy back the content-type
-        for (AsciiString header : relevantHeaders) {
-            request.getHeaders().remove(header);
-            for (String value : finalized.headers().getAll(header)) {
-                request.getHeaders().add(header, value);
+            // if the body is streamed, we have a StreamWriter, otherwise we have a ByteBuf.
+            StreamWriter streamWriter;
+            ByteBuf byteBuf;
+            if (byteBody instanceof AvailableNettyByteBody available) {
+                byteBuf = AvailableNettyByteBody.toByteBuf(available);
+                streamWriter = null;
+            } else {
+                streamWriter = new StreamWriter((StreamingNettyByteBody) byteBody, e -> {
+                    poolHandle.taint();
+                    sink.error(e);
+                });
+                pipeline.addLast(streamWriter);
+                byteBuf = null;
             }
-        }
-        // return the body bytes
+
+            if (log.isTraceEnabled()) {
+                HttpHeadersUtil.trace(log, nettyRequest.headers().names(), nettyRequest.headers()::getAll);
+                if (byteBuf != null) {
+                    traceBody("Request", byteBuf);
+                }
+            }
+
+            pipeline.addLast(ChannelPipelineCustomizer.HANDLER_MICRONAUT_HTTP_RESPONSE, new Http1ResponseHandler(new Http1ResponseHandler.ResponseListener() {
```

Comment:
can this be pulled into an inner class to make it easier to read

Reply:
id prefer not to, it's only like 50 lines but it would require 7 constructor parameters from the method